src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionParsing.scala

```diff
@@ -0,0 +1,224 @@
+package org.broadinstitute.dsde.rawls.expressions
+
+import com.tinkerpop.blueprints._
+import com.tinkerpop.gremlin.java.GremlinPipeline
+import com.tinkerpop.pipes.PipeFunction
+import org.broadinstitute.dsde.rawls.expressions
+import org.broadinstitute.dsde.rawls.model.Workspace
+import scala.collection.JavaConversions._
+import scala.util.{Try, Failure, Success}
+import scala.util.parsing.combinator
+import scala.util.parsing.combinator.JavaTokenParsers
+import scala.util.parsing.input.Reader
+
+object ExpressionTypes {
+  // to make a less verbose type
+  type PipeType = GremlinPipeline[Vertex, Vertex]
+
+  // The output of a single step of the Gremlin pipe. pipeAction is the string representation, for debugging
+  case class PipeResult(result: PipeType, pipeAction: String)
+
+  // types to clarify what a RootFunc takes
+  type WorkspaceName = String
+  type WorkspaceNamespace = String
+  type RootType = String
+  type RootName = String
+
+  // The initial step of a Gremlin pipe.
+  type RootFunc = (PipeType, WorkspaceNamespace, WorkspaceName, RootType, RootName) => PipeResult
+
+  // A single step of the Gremlin pipe.
+  type PipeFunc = PipeType => PipeResult
+
+  // The output of a completed Gremlin pipe; may return >1 result, hence Seq[String]. pipeAction is the step's string representation, for debugging
+  case class FinalResult(result: Seq[Any], pipeAction: String)
+
+  // The terminal step of a Gremlin pipe.
+  type ResultFunc = PipeType => FinalResult
+
+  // a query goes from a root through pipes to the last action, which returns a Seq[String] of processed data from the requested Entities or attributes on entities
+  case class PipelineQuery(rootFunc: RootFunc, pipes: Option[List[PipeFunc]], lastAction: Option[ResultFunc])
+
```

Comment:
Only problem with tuples over case classes if that the former would have to be accessed via either ```_1, _2, ...``` or pattern matching. Which may not be a problem, if you're not accessing them often.